File: src/lib/engine/scoring.ts
// CNC Cut Plan Engine — Scoring, Lookahead, and Cut Position Helpers

import { TreeNode, Piece } from './types';

export function oris(p: Piece): { w: number; h: number }[] {
  if (p.w === p.h) return [{ w: p.w, h: p.h }];
  return [
    { w: p.w, h: p.h },
    { w: p.h, h: p.w },
  ];
}

export function scoreFit(spaceW: number, spaceH: number, pieceW: number, pieceH: number, remaining: Piece[]): number {
  const wasteW = spaceW - pieceW;
  const wasteH = spaceH - pieceH;

  let score = wasteW * spaceH + wasteH * pieceW;

  const pieceArea = pieceW * pieceH;
  score -= pieceArea * 0.5;

  let wFits = false;
  let hFits = false;

  for (const r of remaining) {
    for (const o of oris(r)) {
      if (!wFits && wasteW >= o.w && spaceH >= o.h) wFits = true;
      if (!hFits && pieceW >= o.w && wasteH >= o.h) hFits = true;
      if (wFits && hFits) break;
    }
    if (wFits && hFits) break;
  }

  if (wasteW > 10 && !wFits) score += wasteW * spaceH * 4;
  if (wasteH > 10 && !hFits) score += wasteH * pieceW * 4;

  if (wasteW === 0) score -= spaceH * 20;
  if (wasteH === 0) score -= pieceW * 20;

  return score;
}

export function canResidualFitAnyPiece(
  residualW: number,
  residualH: number,
  remainingPieces: Piece[],
  minBreak: number = 0,
  existingSiblingValues: number[] = [],
  axis: "w" | "h" = "w",
): boolean {
  if (residualW <= 0 || residualH <= 0) return false;
  for (const p of remainingPieces) {
    for (const o of oris(p)) {
      if (o.w <= residualW && o.h <= residualH) {
        if (minBreak > 0 && existingSiblingValues.length > 0) {
          const val = axis === "w" ? o.w : o.h;
          const violates = existingSiblingValues.some((sv) => {
            const diff = Math.abs(sv - val);
            return diff > 0 && diff < minBreak;
          });
          if (violates) continue;
        }
        return true;
      }
    }
  }
  return false;
}

/**
 * Returns true if placing a piece of width `pieceW` in a slot of width `slotW`
 * leaves a residual strip that is narrower than minBreak (but not zero).
 * A zero residual (piece fills the slot exactly) is always valid.
 */
export function zResidualViolatesMinBreak(slotW: number, pieceW: number, minBreak: number): boolean {
  const residual = slotW - pieceW;
  return residual > 0 && residual < minBreak;
}

/** Generic: returns true if newValue creates a gap < minBreak with any sibling value. */
export function siblingViolatesMinBreak(existingValues: number[], newValue: number, minBreak: number): boolean {
  return existingValues.some(v => { const d = Math.abs(v - newValue); return d > 0 && d < minBreak; });
}

export function getZCutPositions(yStrip: TreeNode): number[] {
  const positions: number[] = [];
  let acc = 0;
  for (const z of yStrip.filhos) {
    acc += z.valor * z.multi;
    positions.push(acc);
  }
  return positions;
}

export function getAllZCutPositionsInColumn(colX: TreeNode): number[][] {
  return colX.filhos.map((y) => getZCutPositions(y));
}

export function violatesZMinBreak(
  newCutPositions: number[],
  allPositions: number[][],
  minBreak: number,
  excludeYIndex: number = -1,
): boolean {
  for (let i = 0; i < allPositions.length; i++) {
    if (i === excludeYIndex) continue;
    for (const existPos of allPositions[i]) {
      for (const newPos of newCutPositions) {
        const diff = Math.abs(existPos - newPos);
        if (diff > 0 && diff < minBreak) return true;
      }
    }
  }
  return false;
}

File: wasm-engine/src/scoring.rs
use crate::types::Piece;

/// Both orientations of a piece (deduped if square).
#[inline]
pub fn oris(p: &Piece) -> Vec<(f64, f64)> {
    if (p.w - p.h).abs() < 0.5 {
        vec![(p.w, p.h)]
    } else {
        vec![(p.w, p.h), (p.h, p.w)]
    }
}

pub fn score_fit(space_w: f64, space_h: f64, piece_w: f64, piece_h: f64, remaining: &[Piece]) -> f64 {
    let waste_w = space_w - piece_w;
    let waste_h = space_h - piece_h;
    let mut score = waste_w * space_h + waste_h * piece_w;
    score -= piece_w * piece_h * 0.5;

    let mut w_fits = false;
    let mut h_fits = false;
    for r in remaining {
        for (rw, rh) in oris(r) {
            if !w_fits && waste_w >= rw && space_h >= rh { w_fits = true; }
            if !h_fits && piece_w >= rw && waste_h >= rh { h_fits = true; }
            if w_fits && h_fits { break; }
        }
        if w_fits && h_fits { break; }
    }

    if waste_w > 10.0 && !w_fits { score += waste_w * space_h * 4.0; }
    if waste_h > 10.0 && !h_fits { score += waste_h * piece_w * 4.0; }
    if waste_w == 0.0 { score -= space_h * 20.0; }
    if waste_h == 0.0 { score -= piece_w * 20.0; }
    score
}

pub fn can_residual_fit_any(
    residual_w: f64,
    residual_h: f64,
    remaining: &[Piece],
    min_break: f64,
    existing_sibling_vals: &[f64],
    axis: &str,
) -> bool {
    if residual_w <= 0.0 || residual_h <= 0.0 { return false; }
    for p in remaining {
        for (ow, oh) in oris(p) {
            if ow <= residual_w && oh <= residual_h {
                if min_break > 0.0 && !existing_sibling_vals.is_empty() {
                    let val = if axis == "w" { ow } else { oh };
                    let violates = existing_sibling_vals.iter().any(|&sv| {
                        let diff = (sv - val).abs();
                        diff > 0.0 && diff < min_break
                    });
                    if violates { continue; }
                }
                return true;
            }
        }
    }
    false
}

#[inline]
pub fn z_residual_violates_min_break(slot_w: f64, piece_w: f64, min_break: f64) -> bool {
    let residual = slot_w - piece_w;
    residual > 0.0 && residual < min_break
}

#[inline]
pub fn sibling_violates_min_break(existing: &[f64], new_val: f64, min_break: f64) -> bool {
    existing.iter().any(|&v| {
        let d = (v - new_val).abs();
        d > 0.0 && d < min_break
    })
}

pub fn violates_z_min_break(
    new_cut_positions: &[f64],
    all_positions: &[Vec<f64>],
    min_break: f64,
    exclude_y_index: Option<usize>,
) -> bool {
    for (i, positions) in all_positions.iter().enumerate() {
        if exclude_y_index == Some(i) { continue; }
        for &ep in positions {
            for &np in new_cut_positions {
                let diff = (ep - np).abs();
                if diff > 0.0 && diff < min_break { return true; }
            }
        }
    }
    false
}
